File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=neutral/rep1/test/integration_test.exs
defmodule BrSoccer.IntegrationTest do
  @moduledoc """
  End-to-end tests against the real Kaggle CSVs in `data/kaggle/`.

  These verify the success criteria: all six files load, cross-file queries
  work, and known historical facts come out correctly.
  """
  use ExUnit.Case, async: false

  alias BrSoccer.{Competitions, Loader, Matches, Players, Repo, Stats}

  setup_all do
    Repo.reload()
    :ok
  end

  test "all six CSV files load into matches and players" do
    assert length(Repo.players()) == 18_207
    # Every match source is represented after loading + de-duplication.
    sources = Repo.matches() |> Enum.map(& &1.source) |> Enum.uniq() |> MapSet.new()

    assert MapSet.subset?(
             MapSet.new([:brasileirao_csv, :cup_csv, :libertadores_csv, :br_football, :novo]),
             sources
           )
  end

  test "each individual file is parseable and non-empty" do
    dir = Loader.data_dir()

    for file <- ~w(Brasileirao_Matches.csv Brazilian_Cup_Matches.csv Libertadores_Matches.csv
                   BR-Football-Dataset.csv novo_campeonato_brasileiro.csv fifa_data.csv) do
      rows = BrSoccer.CSV.parse_file(Path.join(dir, file))
      assert length(rows) > 100, "expected data rows in #{file}"
    end
  end

  test "2019 Brasileirão champion is Flamengo with 90 points" do
    table = Competitions.standings(:brasileirao, 2019)
    champ = hd(table)
    assert champ.team == "Flamengo"
    assert champ.points == 90
    assert length(table) == 20
  end

  test "2022 Brasileirão (single-source season) computes a 20-team table" do
    table = Competitions.standings(:brasileirao, 2022)
    assert length(table) == 20
    assert hd(table).position == 1
  end

  test "head-to-head works for a classic rivalry" do
    h = Matches.head_to_head("Flamengo", "Fluminense")
    assert h.total > 20
    assert h.a_wins + h.b_wins + h.draws == h.played
  end

  test "team can be found via name variations across files" do
    # Palmeiras appears as 'Palmeiras', 'Palmeiras-SP' across files; all should resolve.
    comps = Matches.competitions_for("Palmeiras") |> Enum.map(& &1.competition)
    assert :brasileirao in comps
    assert :libertadores in comps
  end

  test "Brazilian players can be filtered and the dataset has the expected scale" do
    brazilians = Players.brazilians()
    assert length(brazilians) == 827
    assert hd(brazilians).nationality == "Brazil"
    # Highest-rated Brazilian should be a world-class name.
    assert hd(brazilians).overall >= 88
  end

  test "cross-file query: Brazilian players grouped by Brazilian clubs" do
    groups = Players.brazilians_at_brazilian_clubs(min_count: 5)
    assert length(groups) > 3
    assert Enum.all?(groups, &(&1.avg_overall > 0))
    # Clubs identified here must exist in the match data.
    keys = Repo.team_keys()
    assert Enum.all?(groups, &MapSet.member?(keys, &1.key))
  end

  test "aggregate stats over the whole Brasileirão are sane" do
    s = Stats.summary(competition: :brasileirao)
    assert s.matches > 4000
    assert s.avg_goals > 2.0 and s.avg_goals < 3.5
    assert s.home_win_rate > 40.0 and s.home_win_rate < 60.0
  end

  test "simple lookups respond well under the 2s budget" do
    {micros, _} = :timer.tc(fn -> Matches.head_to_head("Flamengo", "Corinthians") end)
    assert micros < 2_000_000
  end

  test "aggregate queries respond under the 5s budget" do
    {micros, _} = :timer.tc(fn -> Competitions.standings(:brasileirao, 2019) end)
    assert micros < 5_000_000
  end
end

File: experiments/adrianco/experiment-14/runs/language=typescript_model=opus-4.8-fast_prompt=TDD/rep1/tests/integration.test.ts
import { describe, it, expect, beforeAll } from "vitest";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { loadAll, canonicalMatches } from "../src/loader.js";
import { SoccerDatabase } from "../src/database.js";
import { normalizeTeamName } from "../src/normalize.js";

const here = dirname(fileURLToPath(import.meta.url));
const dataDir = join(here, "..", "data", "kaggle");

let db: SoccerDatabase;

beforeAll(() => {
  db = new SoccerDatabase(loadAll(dataDir));
});

describe("data coverage", () => {
  it("loads all six datasets at their full raw row counts", () => {
    // 4180 + 1337 + 1255 + 10296 + 6886 = 23954 raw match rows.
    const raw = loadAll(dataDir);
    expect(raw.matches.length).toBe(23954);
    expect(raw.players.length).toBe(18207);
  });

  it("canonicalizes overlapping datasets to fewer, non-duplicated matches", () => {
    // Canonical set drops cross-dataset duplication but stays large.
    expect(db.matches.length).toBeLessThan(23954);
    expect(db.matches.length).toBeGreaterThan(15000);
    expect(db.matches.length).toBe(canonicalMatches(loadAll(dataDir).matches).length);
    expect(db.players.length).toBe(18207);
  });

  it("represents every competition", () => {
    const comps = new Set(db.matches.map((m) => m.competition));
    expect(comps.has("Brasileirão Série A")).toBe(true);
    expect(comps.has("Copa do Brasil")).toBe(true);
    expect(comps.has("Copa Libertadores")).toBe(true);
  });
});

describe("real queries", () => {
  it("finds Flamengo vs Fluminense matches with a head-to-head", () => {
    const matches = db.findMatches({ team: "Flamengo", team2: "Fluminense" });
    expect(matches.length).toBeGreaterThan(0);
    const h2h = db.headToHead("Flamengo", "Fluminense");
    expect(h2h.teamAWins + h2h.teamBWins + h2h.draws).toBe(h2h.matches);
  });

  it("computes 2019 Brasileirão standings with Flamengo as champion", () => {
    const table = db.standings("Brasileirão Série A", 2019);
    expect(table.length).toBeGreaterThan(0);
    expect(normalizeTeamName(table[0].team)).toBe("flamengo");
    // Flamengo won the 2019 Brasileirão with 90 points.
    expect(table[0].points).toBe(90);
  });

  it("finds Brazilian players sorted by rating", () => {
    const players = db.findPlayers({ nationality: "Brazil", limit: 5 });
    expect(players.length).toBe(5);
    expect(players[0].overall! >= players[4].overall!).toBe(true);
    expect(players.every((p) => /brazil/i.test(p.nationality))).toBe(true);
  });

  it("finds a known player by name", () => {
    const players = db.findPlayers({ name: "Neymar" });
    expect(players.length).toBeGreaterThan(0);
    expect(players[0].nationality).toBe("Brazil");
  });

  it("computes overall match statistics with a plausible average", () => {
    const stats = db.statistics({});
    expect(stats.totalMatches).toBe(db.matches.length);
    expect(stats.averageGoals).toBeGreaterThan(1.5);
    expect(stats.averageGoals).toBeLessThan(4);
    expect(stats.homeWinRate).toBeGreaterThan(0.3);
  });

  it("groups Brazilian players at Brazilian clubs", () => {
    const groups = db.brazilianPlayersByClub();
    expect(groups.length).toBeGreaterThan(0);
    expect(groups[0].count).toBeGreaterThanOrEqual(groups[groups.length - 1].count);
  });
});

describe("performance", () => {
  it("answers an aggregate query in well under 5 seconds", () => {
    const start = performance.now();
    db.standings("Brasileirão Série A", 2019);
    db.findMatches({ team: "Palmeiras" });
    db.findPlayers({ nationality: "Brazil" });
    const elapsed = performance.now() - start;
    expect(elapsed).toBeLessThan(5000);
  });
});
